sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V1FallbackWriters.scala

```diff
@@ -111,11 +107,10 @@ sealed trait V1FallbackWriters extends SupportsV1Write {
  * A trait that allows Tables that use V1 Writer interfaces to append data.
  */
 trait SupportsV1Write extends SparkPlan {
-  // TODO: We should be able to work on SparkPlans at this point.
-  def plan: LogicalPlan
+  def query: SparkPlan
 
   protected def writeWithV1(relation: InsertableRelation): RDD[InternalRow] = {
-    relation.insert(Dataset.ofRows(sqlContext.sparkSession, plan), overwrite = false)
+    relation.insert(AlreadyPlanned.dataFrame(sqlContext.sparkSession, query), overwrite = false)
```

Comment:
I think `AlreadyPlanned` works well here, as a top-level node. However, there can be problems if we make the framework too general and support using `AlreadyPlanned` as a non-top-levle node.

The physical plan has no stats, so `AlreadyPlanned` has no stats as well. This may change the planning result if the original logical plan has stats, e.g. broadcast join becomes sort merge join.

Reply:
+1. It's probably not worth saving the compilation time if we end up generating a bad plan.